compiler/typecheck/TcDaml.hs

```diff
@@ -13,34 +11,310 @@ import OccName
+      | otherwise
+      -> text "Implementation of method" <+> ppr method <+> text "on interface" <+> ppr target <+> text "failed."
+    Nothing ->
+      vcat [ text "Tried to implement method" <+> ppr method <> text ", but interface" <+> ppr target <+> text "does not have a method with that name."
+           , printListWithHeader
+               empty
+               (text "Method" <+> ppr method <+> text "is only a method on the following interfaces:")
+               (map (ppr . fst) ifaces)
+           ]
+
+data DamlVariant = Template Name | Interface Name | Choice Name
+
+data DamlInfo = DamlInfo
+  { templates :: [Name]
+  , interfaces :: [Name]
+  , choices :: [(Name, Name)]
+  , methods :: [(FastString, (Name, Type))]
+  , implementations :: [(Name, Name)]
+  , views :: [(Name, Type)]
+  }
+
+dedupe :: DamlInfo -> DamlInfo
+dedupe (DamlInfo x0 x1 x2 x3 x4 x5) =
+  DamlInfo
+    (nubSort x0)
+    (nubSort x1)
+    (nubSort x2)
+    (nubSortBy (\(mName, (cName, type_)) -> (mName, cName)) x3)
+    (nubSort x4)
+    (nubSortBy fst x5)
+  where
+    nubSortBy :: Ord b => (a -> b) -> [a] -> [a]
+    nubSortBy f xs = M.elems $ M.fromList $ map (\x -> (f x, x)) xs
+
+isTemplate info name = name `elem` templates info
+isInterface info name = name `elem` interfaces info
+variantName info name
+  | isTemplate info name = text "template" <+> ppr name
+  | isInterface info name = text "interface" <+> ppr name
+  | otherwise = text "type" <+> ppr name
+allImplementedInterfaces info name = [iface | (tpl, iface) <- implementations info, name == tpl]
+allImplementingTemplates info name = [tpl | (tpl, iface) <- implementations info, name == iface]
+implements info tpl iface = (tpl, iface) `elem` implementations info
+choiceImplementor info name = [tplOrIface | (tplOrIface, choice) <- choices info, name == choice]
+interfaceView info name = lookup name (views info)
+definesMethod info method = [nameAndType | (m, nameAndType) <- methods info, m == method]
+
+instance Monoid DamlInfo where
+  mempty = DamlInfo [] [] [] [] [] []
+
+instance Semigroup DamlInfo where
+  (<>) (DamlInfo a0 a1 a2 a3 a4 a5) (DamlInfo b0 b1 b2 b3 b4 b5) =
+    DamlInfo (a0 P.<> b0) (a1 P.<> b1) (a2 P.<> b2) (a3 P.<> b3) (a4 P.<> b4) (a5 P.<> b5)
+
+instance Outputable DamlInfo where
+  ppr info =
+    hang (text "DamlInfo {") 2 $
+    vcat [ text "templates:" <+> ppr (templates info)
+         , text "interfaces:" <+> ppr (interfaces info)
+         , text "choices:" <+> ppr (choices info)
+         , text "methods:" <+> ppr (methods info)
+         , text "implementations:" <+> ppr (implementations info)
+         , text "views:" <+> ppr (views info)
+         ]
+
+extractDamlInfoFromGblEnv :: TcGblEnv -> DamlInfo
+extractDamlInfoFromGblEnv env =
+  foldMap extractDamlInfoFromTyThing (typeEnvElts $ tcg_type_env env) `mappend`
+    foldMap extractDamlInfoFromClsInst (instEnvElts $ tcg_inst_env env)
```

Comment:
did you look into adding this info directly to `TcGblEnv`/`ExternalPackageState` instead of extracting it for each daml error? I'm assuming that's the reason for `tcg_daml_{templates,interfaces,choices}` in `compiler/typecheck/TcRnTypes.hs`. Did you find anything interesting?

Reply:
I fixed that this morning, by coincidence! The extraction code now runs once, and extracted DamlInfo is stored in Env, instead of in TcGblEnv

The `tcg_daml_{templates,interfaces,choices}` fields were just a holdover from a previous iteration - I've removed them as they are now deprecated.